### PR TITLE
Support multi-output in Mir-onX

### PR DESCRIPTION
**Pull request by wmww**
Enhance the `--x11-output` option to support a colon separated list of output sizes. Outputs are lined up horizontally from Mir's perspective, and are each shown in a different X11 window. Useful if you need to test multi-output code paths and do not have auxiliary physical display handy.

### CODE REVIEW

File: src/platforms/mesa/server/x11/X11_resources.cpp [hunk 1]
```diff
@@ -55,3 +58,22 @@ std::shared_ptr<::Display> mx::X11Resources::get_conn()
...
+{
+    output_configs[win] = configuration;
+}
+
+void mx::X11Resources::clear_output_config_for_win(Window win)
+{
+    output_configs.erase(win);
+}
+
+std::experimental::optional<mg::DisplayConfigurationOutput*> mx::X11Resources::get_output_config_for_win(Window win)
```
Comment (RAOF) [inline src/platforms/mesa/server/x11/X11_resources.cpp]: If you're going to return a `std::experimental::optional`, I'd mildly prefer `std::experimental::optional<mg::DisplayConfigurationOutput&>` to the pointer version.

File: src/platforms/mesa/server/x11/graphics/display.cpp [hunk 2]
```diff
@@ -231,39 +234,50 @@ unsigned long mgx::X11Window::red_mask() const
...
-                         *gl_config);
+    for (auto const& requested_size : requested_sizes)
+    {
+        auto actual_size = clip_to_display(x_dpy, requested_size);
+        auto window = std::make_unique<X11Window>(x_dpy, shared_egl.display(), actual_size, shared_egl.config());
+        auto red_mask = window->red_mask();
+        auto pf = (red_mask == 0xFF0000 ?
+            mir_pixel_format_argb_8888 :
+            mir_pixel_format_abgr_8888);
+        auto configuration = static_cast<std::unique_ptr<graphics::DisplayConfigurationOutput>>(
```
[the remark this answers is not in the document]
Comment (wmww) [reply src/platforms/mesa/server/x11/graphics/display.cpp]: Oops. I ended up with an unnecessary `graphics::X::DisplayConfigurationOutput` subclass that needed to be cast from. Refactored that into a static builder method `graphics::X::DisplayConfiguration`, which cleans this up.

File: src/platforms/mesa/server/x11/X11_resources.cpp (hunk 1 above)
Comment (wmww) [reply src/platforms/mesa/server/x11/X11_resources.cpp]: I don't think that's allowed. When I try, I get
```
/home/wmww/code/mir/src/platforms/mesa/server/x11/X11_resources.cpp:72:122:   required from here
/usr/include/c++/8/experimental/optional:461:15: error: non-static data member ‘std::experimental::fundamentals_v1::_Optional_base<const mir::graphics::DisplayConfigurationOutput&, false>::<unnamed union>::_M_payload’ in a union may not have reference type ‘std::experimental::fundamentals_v1::_Optional_base<const mir::graphics::DisplayConfigurationOutput&, false>::_Stored_type’ {aka ‘const mir::graphics::DisplayConfigurationOutput&’}
```
I also found [this](https://foonathan.net/blog/2018/07/12/optional-reference.html) on the topic.

**Comment by wmww** [comment]
Added tests and cleared up the ownership issues (hopefully). The reason I use a global map is because events get a `Window`, and I didn't find a way to attach use data onto that. Now the values windows map to are const weak pointers, so maybe better?

**Comment by AlanGriffiths** [comment]
The layout of the X windows is a bit odd:

- The first window is placed top left of the *current* output.
- Each subsequent window is placed to the right of the previous window **unless** it would extend outside the *first* output, in which case it is placed on top of the previous window.

So, if I'm working on my right monitor, all the windows pile up on top of each other.
If I'm working on my left monitor, the first couple of windows tile nicely and the rest pile up on top of each other.

This isn't a blocker, but it would make more sense to use the entire display area.

**Comment by wmww** [comment]
@AlanGriffiths are you talking about how the "output" windows are placed within the parent desktop environment? If so, I don't believe we're exerting any control of that client-side (I certainly didn't touch any window moving code). X obviously gives us enough control to place windows how we like, but that would be an entirely new feature (which I imagine we'd agree isn't worth it).

**Comment by AlanGriffiths** [comment]
> @AlanGriffiths are you talking about how the "output" windows are placed within the parent desktop environment? If so, I don't believe we're exerting any control of that client-side (I certainly didn't touch any window moving code). X obviously gives us enough control to place windows how we like, but that would be an entirely new feature (which I imagine we'd agree isn't worth it).

Yeah, I didn't notice any code for it, but the different behaviour depending on different outputs made me wonder. (As I said, "non-blocking".)

**Comment by RAOF** [comment]
More obvious ownership, *and* tests? Be still my beating heart!

I don't *think* we're doing anything that would make it less likely a decent window manager would position our windows well. GNOME Shell should be doing a better job, here.

bors r+

**Pull request merged**